### PR TITLE
Ballot verification, secret sharing, and verifiable decryption

### PR DESCRIPTION
**Pull request by tschudid**
# Purpose

Implement ballot verification, secret sharing of guardian keys, and verifiable decryption.

# Changes

This PR implements the following changes:

- Guardian Key Generation:
  - Add Schnorr proofs for public keys.
  - Add secret sharing of secret keys.
- Ballot encryption:
  - Add missing pieces to verification of range proofs.
  - Add verification of encrypted ballots.
- Ballot Aggregation:
  - Add the option to scale ballots with public weights.
- Verifiable decryption:
  - Add decryption functionality.
  - Add proof of correct decryption.
- Utils:
  - Add an abstraction for field and group elements. This bundles integer arithmetics into one place. In particular, one can now easily replace the big-integer library if desired.

### CODE REVIEW

**Comment by tschudid** [comment]
@microsoft-github-policy-service agree company="Concordium"

**Comment by marshrayms** [comment]
Thank you for this contribution!
It will take me awhile to go through it, but it looks great.

**Comment by marshrayms** [comment]
This is not forgotten, thank you for continuing to maintain the branch.
:- )

File: src/eg/Cargo.toml [hunk 1]
```diff
@@ -7,15 +7,18 @@ publish = false
...
 num-bigint = "0.4"
 num-traits = "0.2"
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 sha2 = "0.10"
 static_assertions = "1.1.0"
+thiserror = "1.0"
 util = { path = "../util" }
 base64 = "0.21.2"
+chrono = { version = "0.4.34", features = ["serde"] }
```
Comment (marshrayms) [inline src/eg/Cargo.toml:21]: Earlier versions of EGDS 2.0 specified a date, but after discussions we softened that language due to the crazy complexity datetimes bring in a specification document. As much as I'm a fan of `chrono`, I'd really prefer that the core `eg` crate not depend on any date/time library.
We'll take it in this PR, but let's make a plan to move it to the 'info' string soon.

File: src/eg/src/vec1.rs [hunk 2]
```diff
@@ -10,7 +10,19 @@ use std::collections::TryReserveError;
 use anyhow::{ensure, Error, Result};
 use serde::{Deserialize, Deserializer, Serialize, Serializer};
 
-use crate::index::Index;
+use crate::{index::Index, zk::ProofRange};
+
+/// Trait for specifying a type that a `Vec1<T>` should be indexed by.
+pub trait HasIndexType {
```
Comment (marshrayms) [inline src/eg/src/vec1.rs:16]: Nice!

File: src/eg/src/ballot.rs [hunk 3]
```diff
@@ -29,8 +37,11 @@ pub enum BallotState {
 /// An encrypted ballot.
 #[derive(Debug, Serialize, Deserialize)]
 pub struct BallotEncrypted {
+    /// The index of ballot style that this ballot belongs to.
+    pub ballot_style_index: BallotStyleIndex,
+
     /// Contests in this ballot
-    pub contests: Vec1<ContestEncrypted>,
+    pub contests: BTreeMap<ContestIndex, ContestEncrypted>,
```
Comment (marshrayms) [inline src/eg/src/ballot.rs:44]: This is a design change worth thinking about carefully. My thinking was that the `contests` vec must have exactly the number of elements as indicated by the ballot style referred to by its `ballot_style_index`.
Allowing every ballot to specify its own set of contests introduces the possibility of a subtle inconsistency: a ballot could specify votes for a contest not present in its ballot style. Indeed, the `ContestNotInManifest` error condition was defined, as would be appropriate for validating a ballot style definition. But invalid ballots could produce `BallotStyleNotInManifest` and/or `ContestNotInBallotStyle` errors.

**Review by marshrayms — APPROVED** [review]

**Pull request merged**